sql/src/main/java/io/crate/action/job/ContextPreparer.java

```diff
@@ -88,65 +97,207 @@ public ContextPreparer(MapSideDataCollectOperation collectOperation,
         innerPreparer = new InnerPreparer();
     }
 
-    public void prepare(UUID jobId,
-                        NodeOperation nodeOperation,
-                        JobExecutionContext.Builder contextBuilder,
-                        @Nullable RowReceiver rowReceiver) {
-        PreparerContext preparerContext = new PreparerContext(jobId, nodeOperation, rowReceiver);
-        ExecutionSubContext subContext = innerPreparer.process(nodeOperation.executionPhase(), preparerContext);
-        contextBuilder.addSubContext(nodeOperation.executionPhase().executionPhaseId(), subContext);
+    public List<ListenableFuture<Bucket>> prepareOnRemote(UUID jobId,
+                                                          Iterable<? extends NodeOperation> nodeOperations,
+                                                          JobExecutionContext.Builder contextBuilder) {
+        PreparerContext preparerContext = new PreparerContext(jobId);
```

Comment:
can you change the PreparerContext constructor to take the RowDownstreamFactory?